src/pal/pal_apple.h

```diff
@@ -1,92 +1,218 @@
+     * "category 3", thereby affording us a "well-known" memory tag that can be
+     * easily identified in tools such as vmmap(1) or Instruments.
      *
-     * Darwin platform allows to gives an ID to anonymous pages via
-     * the VM_MAKE_TAG's macro, from 240 up to 255 are guaranteed
-     * to be free of usage, however eventually a lower could be taken
-     * (e.g. LLVM sanitizers has 99) so we can monitor their states
-     * via vmmap for instance. This value is provided to `mmap` as the file
-     * descriptor for the mapping.
      */
-    static constexpr int anonymous_memory_fd = VM_MAKE_TAG(PALAnonID);
 
-    /**
-     * Note: The root's implementation works fine on Intel
-     * however mprotect/PROT_NONE fails on ARM
-     * especially since the 11.2 release (seems known issue
-     * spotted in various projects; might be a temporary fix).
-     */
-    template<bool page_aligned = false>
-    static void zero(void* p, size_t size) noexcept
-    {
-      if constexpr (Aal::aal_name != ARM)
-        PALBSD::zero(p, size);
-      else
-        ::bzero(p, size);
-    }
+    // Encoded memory tag passed to `mmap`.
+    static constexpr int anonymous_memory_fd =
+      int(VM_MAKE_TAG(uint32_t(PALAnonID)));
+
+    // Encoded memory tag passed to `mach_vm_map`.
+    static constexpr int default_mach_vm_map_flags =
+      int(VM_MAKE_TAG(uint32_t(PALAnonID)));
 
-#  if defined(PLATFORM_IS_ARM)
     /**
-     * Overriding here to mark the page as reusable
-     * rolling it as much as necessary.
-     * As above, the x86 h/w worked alright without this change
-     * however now large allocations work better and more reliably
-     * with on ARM, not to mention better RSS number accuracy
-     * for tools based on task_info API.
+     * Notify platform that we will not be using these pages.
      */
     static void notify_not_using(void* p, size_t size) noexcept
     {
-      SNMALLOC_ASSERT(is_aligned_block<PALBSD::page_size>(p, size));
-#    ifdef USE_POSIX_COMMIT_CHECKS
+      SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
       memset(p, 0x5a, size);
-#    endif
+#  endif
+
+      // `MADV_FREE_REUSABLE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      //
+      // macOS 11 Big Sur may behave in an undocumented manner.
       while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
         ;
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // This must occur after `MADV_FREE_REUSABLE`.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_NONE);
+#  endif
     }
 
     /**
-     * same remark as above but we need to mark the page as REUSE
-     * first
+     * Notify platform that we will be using these pages.
      */
     template<ZeroMem zero_mem>
     static void notify_using(void* p, size_t size) noexcept
```

Comment:
This function now looks almost identical to the POSIX PAL version.  Why do we need a separate one?

Reply:
`MADV_FREE` != `MADV_FREE_REUSABLE` on XNU.
Please see the above comment. https://github.com/microsoft/snmalloc/pull/336#discussion_r655723362

Reply:
Thanks.  We already have some template parameters for the generic PALs that allow tweaking things like the default `mmap` flags, I was wondering why this couldn't be done to make `MADV_FREE` something tunable, but it looks as if the order of operations is sufficiently different that this isn't possible (or, rather, would add a lot more complexity than having a custom implementation here).